impeller/typographer/backends/stb/text_render_context_stb.cc

```diff
@@ -0,0 +1,587 @@
+    }
+    DrawGlyph(bitmap.get(), pair, pos.value(), has_color);
+  }
+  return true;
+}
+
+static std::shared_ptr<STBBitmap> CreateAtlasBitmap(const GlyphAtlas& atlas,
+                                                    const ISize& atlas_size) {
+  TRACE_EVENT0("impeller", __FUNCTION__);
+
+  size_t bytes_per_pixel = 1;
+  if (atlas.GetType() == GlyphAtlas::Type::kColorBitmap &&
+      !DISABLE_COLOR_FONT_SUPPORT) {
+    bytes_per_pixel = COLOR_FONT_BPP;
+  }
+  auto bitmap = std::make_shared<STBBitmap>(atlas_size.width, atlas_size.height,
+                                            bytes_per_pixel);
+
+  bool has_color = atlas.GetType() == GlyphAtlas::Type::kColorBitmap;
+
+  atlas.IterateGlyphs([&bitmap, has_color](const FontGlyphPair& font_glyph,
+                                           const Rect& location) -> bool {
+    DrawGlyph(bitmap.get(), font_glyph, location, has_color);
+    return true;
+  });
+
+  return bitmap;
+}
+
+// static bool UpdateGlyphTextureAtlas(std::shared_ptr<SkBitmap> bitmap,
+static bool UpdateGlyphTextureAtlas(std::shared_ptr<STBBitmap>& bitmap,
+                                    const std::shared_ptr<Texture>& texture) {
+  TRACE_EVENT0("impeller", __FUNCTION__);
+
+  FML_DCHECK(bitmap != nullptr);
+
+  auto texture_descriptor = texture->GetTextureDescriptor();
+
+  auto mapping = std::make_shared<fml::NonOwnedMapping>(
+      reinterpret_cast<const uint8_t*>(bitmap->GetPixels()),  // data
+      texture_descriptor.GetByteSizeOfBaseMipLevel()          // size
+      // As the bitmap is static in this module I believe we don't need to
+      // specify a release proc.
+  );
+
+  return texture->SetContents(mapping);
+}
+
+static std::shared_ptr<Texture> UploadGlyphTextureAtlas(
+    const std::shared_ptr<Allocator>& allocator,
+    std::shared_ptr<STBBitmap>& bitmap,
+    const ISize& atlas_size,
+    PixelFormat format) {
+  TRACE_EVENT0("impeller", __FUNCTION__);
+  if (!allocator) {
+    return nullptr;
+  }
+
+  FML_DCHECK(bitmap != nullptr);
+
+  TextureDescriptor texture_descriptor;
+  texture_descriptor.storage_mode = StorageMode::kHostVisible;
+  texture_descriptor.format = format;
+  texture_descriptor.size = atlas_size;
+
+  if (bitmap->GetRowBytes() * bitmap->GetHeight() !=
+      texture_descriptor.GetByteSizeOfBaseMipLevel()) {
+    return nullptr;
+  }
+
+  auto texture = allocator->CreateTexture(texture_descriptor);
+  if (!texture || !texture->IsValid()) {
+    return nullptr;
+  }
+  texture->SetLabel("GlyphAtlas");
+
+  auto mapping = std::make_shared<fml::NonOwnedMapping>(
+      reinterpret_cast<const uint8_t*>(bitmap->GetPixels()),  // data
+      texture_descriptor.GetByteSizeOfBaseMipLevel()          // size
+      // As the bitmap is static in this module I believe we don't need to
+      // specify a release proc.
+  );
+
+  if (!texture->SetContents(mapping)) {
+    return nullptr;
+  }
+  return texture;
+}
+
+std::shared_ptr<GlyphAtlas> TextRenderContextSTB::CreateGlyphAtlas(
```

Comment:
In a later patch, we can dry this up with the Skia typographer context. Can we file a followup issue?

Reply:
True, especially now that these globals aren't floating around. https://github.com/flutter/flutter/issues/133029